### PR TITLE
jwt_authn: Set metadata irrespective of success/failure of JWT Verification

### LINKED ISSUE #32813: Cannot access JWT provider failed_status_in_metadata in lua response
*Title*: Cannot access JWT provider failed_status_in_metadata in lua response

*Description*:

Using recent envoy build:

`"version": "1ddf0f5d8fade9c4209a12b34f6959bc7a90e840/1.30.0-dev/Clean/RELEASE/BoringSSL",`

When setting `failed_status_in_metadata: failed_status` in a jwt_authn filter provider and supplying an invalid JWT token in a request, no dynamic metadata is available in envoy.filters.http.jwt_authn.failed_status` within the lua envoy_on_response() function.

*Repro steps*:

Configure the jwt_authn plugin and lua plugins as follows:

```yaml
- name: envoy.filters.http.jwt_authn
  typed_config:
    "@type": type.googleapis.com/envoy.extensions.filters.http.jwt_authn.v3.JwtAuthentication
    providers:
      authspa:
        issuer: https://issuer
        remote_jwks:
          http_uri:
            uri: https://jwks_uri
            cluster: authspa_jwks_cluster
            timeout: 1s
          cache_duration:
            seconds: 300
        forward: true
        forward_payload_header: x-jwt
        payload_in_metadata: jwt_payload
        failed_status_in_metadata: failed_status
    rules:
    - match:
        prefix: /
      requires:
        provider_name: authspa
- name: envoy.filters.http.lua
   typed_config:
     "@type": type.googleapis.com/envoy.extensions.filters.http.lua.v3.Lua
   default_source_code:
     inline_string: |
       function envoy_on_response(response_handle)
         local meta = response_handle:streamInfo():dynamicMetadata()
         local jwtMeta = meta:get("envoy.filters.http.jwt_authn")
         if type(jwtMeta) == "nil"
         then
           response_handle:logWarn("envoy.filters.http.jwt_authn dynamic metadata is not available.")
           return
         end
```

*Logs*:

```
[2024-03-10 10:51:40.220][40][debug][jwt] [source/extensions/filters/http/jwt_authn/authenticator.cc:162] authspa: JWT authentication starts (allow_failed=false), tokens size=1
[2024-03-10 10:51:40.220][40][debug][jwt] [source/extensions/filters/http/jwt_authn/authenticator.cc:450] Code: 3 Message: Jwt is expired
[2024-03-10 10:51:40.220][40][debug][http] [source/common/http/filter_manager.cc:1017] [Tags: "ConnectionId":"0","StreamId":"7547608061384460992"] Preparing local reply with details jwt_authn_access_denied{Jwt_is_expired}
[2024-03-10 10:51:40.220][40][debug][http] [source/common/http/filter_manager.cc:1059] [Tags: "ConnectionId":"0","StreamId":"7547608061384460992"] Executing sending local reply.
[2024-03-11 10:34:43.196][37][debug][jwt] [source/extensions/filters/http/jwt_authn/filter.cc:97] Called Filter : decodeHeaders Stop
[2024-03-10 10:51:40.220][40][warning][lua] [source/extensions/filters/http/lua/lua_filter.cc:925] script log: envoy.filters.http.jwt_authn dynamic metadata is not available.
```

### PR DESCRIPTION
**Pull request by arulthileeban**
Commit Message: jwt_authn: Set metadata irrespective of success/failure of JWT Verification

Previously, metadata was only set for successful JWT verification, restricting "failed_status_in_metadata". This change removes the condition, allowing both "payload_in_metadata" and "failed_status_in_metadata" to be set as needed.

If this condition was added to restrict setting "payload_in_metadata" only on successful verification, it is unnecessary since "payload_in_metadata" is set only by AuthenticatorImpl:handleGoodJwt().

Risk Level: Low
Testing: Unit tests
Docs Changes: N/A
Release Notes:
Platform Specific Features: N/A
Fixes #32813

### CODE REVIEW

**Comment by adisuissa** [comment]
@TAOXUY can you PTAL?

**Comment by tyxia** [comment]
@kyessenov  is OOO and probably won't be able to look at this PR.

@TAOXUY Friendly ping on this as you are the code-owner. I will pass it onto a maintainer review once it is LGTY. Thanks!

**Comment by tyxia** [comment]
/wait-any

@dio @qiwzhang   Looks like you touched this feature in #18140 and #4707. The logic here has been `successfully verified JWT` .  wondering what is your opinions on this PR? Thanks in advance!

**Comment by arulthileeban** [comment]
@tyxia Anything that can be done from my end to help get this PR going? It has been stuck for quite a while. 

If you are able to add the test I've added, to the upstream code, the setExtractedData callback will fail which is what sets the failed_status_in_metadata. This change is required to make this config work. Also, the "successfully verified JWT" logic was likely there earlier since metadata was only set for successful JWTs through "payload_in_metadata". The failed_status_in_metadata logic came later and this check was probably overlooked when adding that. The current tests also miss this. 

The payload_in_metadata is still set only by [handleGoodJWT](https://github.com/envoyproxy/envoy/blob/46f86da8a471e6a23ce5cf1431cc9abce2dc260e/source/extensions/filters/http/jwt_authn/authenticator.cc#L414) which makes this check redundant. 

HandleGoodJWT again is only called when a JWT cache hit [happens](https://github.com/envoyproxy/envoy/blob/46f86da8a471e6a23ce5cf1431cc9abce2dc260e/source/extensions/filters/http/jwt_authn/authenticator.cc#L264) or if verification [succeeds](https://github.com/envoyproxy/envoy/blob/46f86da8a471e6a23ce5cf1431cc9abce2dc260e/source/extensions/filters/http/jwt_authn/authenticator.cc#L322)

**Review by tyxia — COMMENTED** [review]
Thanks for detailed information @arulthileeban ! 

This CL LGTM, but I am not familiar with this extension so I think will need the extension owner/other maintainer's approval

@lizan Would you get a chance to look at this PR. Also cced @danieldradware (who added failed_status_to_metadata initially) and @qiwzhang(who reviewed the initial PR)

**Comment by arulthileeban** [comment]
@TAOXUY @lizan Could you PTAL?

**Review by tyxia — APPROVED** [review]
LGTM

I think it is also good to have a pass from code owner/existing users

**Comment by arulthileeban** [comment]
Would it be possible to get this fix into the upcoming release?

**Review by kyessenov — APPROVED** [review]
This makes sense to me. I'm not an extension owner, but I've been looking in depth into it recently.

**Pull request merged**
